dagger/manifest.nim

```diff
@@ -0,0 +1,197 @@
+## Nim-Dagger
+## Copyright (c) 2021 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+{.push raises: [Defect].}
+
+import pkg/libp2p
+import pkg/libp2p/protobuf/minprotobuf
+import pkg/questionable
+import pkg/questionable/results
+import pkg/chronicles
+import pkg/chronos
+
+import ./blocktype
+
+const
+  ManifestCodec* = multiCodec("dag-pb")
+
+type
+  BlocksManifest* = object
+    blocks: seq[Cid]
+    htree: ?Cid
+    version*: CidVersion
+    hcodec*: MultiCodec
+    codec*: MultiCodec
+    cidEmptyDigest: MultiHash
+
+proc len*(b: BlocksManifest): int = b.blocks.len
+
+iterator items*(b: BlocksManifest): Cid =
+  for b in b.blocks:
+    yield b
+
+proc hashBytes(mh: MultiHash): seq[byte] =
+  mh.data.buffer[mh.dpos..(mh.dpos + mh.size - 1)]
+
+proc cid*(b: var BlocksManifest): ?Cid =
+  if b.htree.isSome:
+    return b.htree
+
+  var
+    stack: seq[MultiHash]
+
+  if stack.len == 1:
+    stack.add(b.cidEmptyDigest)
+
+  for cid in b.blocks:
+    if mhash =? cid.mhash:
+      stack.add(mhash)
+
+    while stack.len > 1:
+      var
+        (b1, b2) = (stack.pop(), stack.pop())
+
+      var
+        digest = MultiHash.digest(
+          $b.hcodec,
+          (b1.hashBytes() & b2.hashBytes()))
+
+      without mh =? digest:
+        return Cid.none
```

Comment:
The problem with `?` is that the `Result` types should match, which isn't the case here since `Cid`, `Multihash`, etc, all have their own error types.

Reply:
But maybe we can return `Result` instead of `Option`? I'm either way on it, it just didn't seem like it would be relevant to have an error here - you either generated the `Cid` based on the `Tree Hash` or you didn't?

Reply:
Yeah, it's a bit verbose but it would be better to return a Result and do something like
```nim
let mh = ? MultiHash.digest(xx).mapErr(x => $x) # or any other smart logic in mapErr
```